### DISCUSSION TITLE
TraceQL structural operator how to combine/select fields from e.g. parent

### DISCUSSION THREAD

**Discussion by WizzardMaker** [grafana/tempo#4313] [Q&A]
We want to aggregate run time metrics of LLM calls in an application with categorized searches.

Each LLM call has a span associated with all its metrics and that span is a child of a "category" span (with other child spans as well)
Now the category span has an attribute which we want as "group by" key for the aggregation, and a single trace contains multiple category spans.

Problem is, that we don't really see a performant/easy way to get the tag from the parent span in a child query

The query looks something like this:
```
{resource.service.name="integration-backend" && name="completion_method"} > {name="AzureChatOpenAI.chat"} |  select(span.tag, span.gen_ai.prompt.0.content, span.gen_ai.usage.prompt_tokens, span.gen_ai.usage.completion_tokens)
```

But that won't get us the span.tag of the parent

A way that could work would be to use a union structural, but parent and child would need to have a common value to aggregate them later with e.g. Grafana transformations. We couldn't find any attribute that "links" both together.
Example result for a union looks like this (there are more rows in the result than in the image):
![image](https://github.com/user-attachments/assets/89acbebe-12d1-4fc5-b697-9a1a1b4846f5)

Our goal is to have a query that results in a single row combining parent-child 
What would be the best way to accomplish a true "union" of parent and child span attributes in the same result row?

**Comment by mapno** [comment]
Hi @WizzardMaker. What visualization are you using: explore or a table panel? With a [union structural operator](https://grafana.com/docs/tempo/latest/traceql/#union-structural) and selecting `trace:id` you should be able to associate spans belonging to the same trace.

**Reply by joe-elliott** [reply]
union structural operators will return spans from both sides of the operator. so you will get both parents and children in the resultset but they won't be "joined" on the same row.

you may be able to do some grafana table transformation shenanigans to join them up?

**Reply by WizzardMaker** [reply]
The problem with that is, that there are multiple category spans in a single trace, so joining by trace:id wouldn't really work in this case sadly

Mock-example trace:
```
trace:
- category span: tag=A
  - ai span: metric=0
- (...irrelevant spans...)
- category span: tag=X
  - ai span: metric=15

(The category spans can also appear at any depth in the trace)
```


The desired table would then be:
| tag | (avg) metric |
| - | - |
| A | 0 |
| B | 15 |
